enterprise/coderd/workspaceproxy.go

```diff
@@ -137,3 +157,126 @@ func convertProxy(p database.WorkspaceProxy) codersdk.WorkspaceProxy {
+			if len(parts) != 2 {
+				httpapi.Write(ctx, w, http.StatusUnauthorized, codersdk.Response{
+					Message: "Invalid external proxy token",
+				})
+				return
+			}
+			proxyID, err := uuid.Parse(parts[0])
+			if err != nil {
+				httpapi.Write(ctx, w, http.StatusUnauthorized, codersdk.Response{
+					Message: "Invalid external proxy token",
+				})
+				return
+			}
+			secret := parts[1]
+			if len(secret) != 64 {
+				httpapi.Write(ctx, w, http.StatusUnauthorized, codersdk.Response{
+					Message: "Invalid external proxy token",
+				})
+				return
+			}
+
+			// Get the proxy.
+			proxy, err := db.GetWorkspaceProxyByID(ctx, proxyID)
+			if xerrors.Is(err, sql.ErrNoRows) {
+				// Proxy IDs are public so we don't care about leaking them via
+				// timing attacks.
+				httpapi.Write(ctx, w, http.StatusUnauthorized, codersdk.Response{
+					Message: "Invalid external proxy token",
+					Detail:  "Proxy not found.",
+				})
+				return
+			}
+			if err != nil {
+				httpapi.InternalServerError(w, err)
+				return
+			}
+			if proxy.Deleted {
+				httpapi.Write(ctx, w, http.StatusUnauthorized, codersdk.Response{
+					Message: "Invalid external proxy token",
+					Detail:  "Proxy has been deleted.",
+				})
+				return
+			}
+
+			// Do a subtle constant time comparison of the hash of the secret.
+			hashedSecret := sha256.Sum256([]byte(secret))
+			if subtle.ConstantTimeCompare(proxy.TokenHashedSecret, hashedSecret[:]) != 1 {
+				httpapi.Write(ctx, w, http.StatusUnauthorized, codersdk.Response{
+					Message: "Invalid external proxy token",
+					Detail:  "Invalid proxy token secret.",
+				})
+				return
+			}
+
+			// TODO: set on context.
+
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+// @Summary Issue signed workspace app token
+// @ID issue-signed-workspace-app-token
+// @Security CoderSessionToken
+// @Accept json
+// @Produce json
+// @Tags Enterprise
+// @Param request body proxysdk.IssueSignedAppTokenRequest true "Issue signed app token request"
+// @Success 201 {object} proxysdk.IssueSignedAppTokenResponse
+// @Router /proxy-internal/issue-signed-app-token [post]
+// @x-apidocgen {"skip": true}
+func (api *API) issueSignedAppToken(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	// NOTE: this endpoint will return JSON on success, but will (usually)
+	// return a self-contained HTML error page on failure. The external proxy
+	// should forward any non-201 response to the client.
+
+	var req proxysdk.IssueSignedAppTokenRequest
+	if !httpapi.Read(ctx, rw, r, &req) {
+		return
+	}
+
+	// HACK: the CreateToken code reads the session token from the request, but
+	// since the session token is in a header, we need to make a fake request.
+	//
+	// TODO(@dean): fix this hack. This could be fixed by providing the token as
+	// a param to CreateToken instead of the whole request.
+	fakeReq := httptest.NewRequest("GET", req.AppRequest.BasePath, nil)
+	fakeReq.Header.Set(codersdk.SessionTokenHeader, req.SessionToken)
```

Comment:
I think this is actually ok. We are essentially recreating the request as if we were the acting user.